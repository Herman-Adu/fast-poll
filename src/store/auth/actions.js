export const SIGN_IN = 'auth/SIGN_IN'
export const SIGN_OUT = 'auth/SIGN_OUT'
export const LOADING = 'auth/LOADING'

export const signInAction = ({ uid, isAnonymous, photoURL }) => ({
  type: SIGN_IN,
  payload: {
    uid,
    isAnonymous,
    photoURL,
  }
})

export const signOutAction = () => ({
  type: SIGN_OUT,
  payload: null
})

export const loadingAction = (isLoading) => ({
  type: LOADING,
  payload: {
    isLoading
  }
})

export const signIn = (auth, provider) => (dispatch) => {
  dispatch(loadingAction(true))
  switch (provider) {
    // the auth listener will handle the success cases
    case 'google':
      return auth()
        .signInWithPopup(new auth.GoogleAuthProvider())
        .then(() => {
          dispatch(loadingAction(false))
        })
        .catch(error => {
          // eslint-disable-next-line no-console
          console.error(error);
          // TODO: notify the user of the error
          return error;
        });

    case 'anonymous':
      return auth()
        .signInAnonymously()
        .then(() => {
          dispatch(loadingAction(false))
        })
        .catch(error => {
          // eslint-disable-next-line no-console
          console.error(error);
          // TODO: notify the user of the error
          return error;
        });

    default:
      const reason = 'Invalid provider passed to signIn method';
      // eslint-disable-next-line no-console
      console.error(reason);
      return Promise.reject(reason);
  }
}

export const signOut = (auth) => (dispatch) => {
  return auth().signOut().then(() => {
    dispatch(signOutAction())
  })
}

export const startListeningToAuthChanges = (auth) => (dispatch) => {
  return auth().onAuthStateChanged(user => {
    if (user) {
      // if user exists sign-in!
      dispatch(signInAction(user))
    } else {
      // otherwise sign-out!
      dispatch(signOutAction())
    }
  });
}