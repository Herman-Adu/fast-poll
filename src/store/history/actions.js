export const LOADING = 'history/LOADING';
export const ADD_POLLS = 'history/ADD_POLLS';

export const loadingAction = isLoading => ({
  type: LOADING,
  payload: {
    isLoading,
  },
});

export const addPollsAction = (type, polls) => ({
  type: ADD_POLLS,
  payload: {
    type,
    polls,
  },
});

export const getPollHistory = (firebase, { uid }) => dispatch => {
  const LIMIT = 12;
  const userRef = firebase.users.doc(uid);

  dispatch(loadingAction(true));

  return Promise.all([
    userRef
      .collection('createdPolls')
      .orderBy('date')
      .limit(LIMIT)
      .get(),
    userRef
      .collection('votedOnPolls')
      .orderBy('date')
      .limit(LIMIT)
      .get(),
  ])
    .then(([created, votedOn]) => {
      const createdPolls = [];
      const votedOnPolls = [];

      created.forEach(doc => {
        createdPolls.push(doc.data());
      });
      votedOn.forEach(doc => {
        votedOnPolls.push(doc.data());
      });

      dispatch(loadingAction(false));
      dispatch(addPollsAction('created', createdPolls));
      dispatch(addPollsAction('votedOn', votedOnPolls));

      return [
        created.docs[created.docs.length - 1],
        votedOn.docs[votedOn.docs.length - 1],
      ];
    })
    .catch(error => {
      // eslint-disable-next-line no-console
      console.error(error);
      // TODO: notify the user of the error
    });
};
