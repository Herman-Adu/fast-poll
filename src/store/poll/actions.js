export const LOADING = 'poll/LOADING'
export const UPDATE_TITLE = 'poll/UPDATE_TITLE'
export const UPDATE_SELECTION = 'poll/UPDATE_SELECTION'
export const UPDATE_VOTED = 'poll/UPDATE_VOTED'

export const loadingAction = (isLoading) => ({
  type: LOADING,
  payload: {
    isLoading
  }
})

export const updateTitleAction = (title) => ({
  type: UPDATE_TITLE,
  payload: {
    title
  }
})

export const updateSelectionAction = (selection) => ({
  type: UPDATE_SELECTION,
  payload: {
    selection
  }
})

export const updateVotedAction = (hasVoted) => ({
  type: UPDATE_VOTED,
  payload: {
    hasVoted
  }
})

export const updateTitle = (title) => (dispatch) => {
  return dispatch(updateTitleAction(title))
}

export const createPoll = (firebase, history, data) => dispatch => {
  const { pollId, options, title } = data

  dispatch(loadingAction(true))

  return firebase.polls
    .doc(pollId)
    .set({
      title,
      id: pollId,
      options: options.map(({ text, id }) => ({ text, optionId: id })),
    })
    .then(() => {
      dispatch(loadingAction(true))

      history.push(`/poll/${pollId}`);
    })
    .catch(error => {
      // eslint-disable-next-line no-console
      console.error(error);
      // TODO: notify the user of the error
    });
}