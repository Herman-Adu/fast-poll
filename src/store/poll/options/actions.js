export const ADD_OPTION = 'poll/options/ADD_OPTION'
export const REMOVE_OPTION = 'poll/options/REMOVE_OPTION'
export const UPDATE_OPTION = 'poll/options/UPDATE_OPTION'
export const UPDATE_OPTION_ORDER = 'poll/options/UPDATE_OPTION_ORDER'

export const addOptionAction = (id) => ({
  type: ADD_OPTION,
  payload: {
    id,
  }
})

export const removeOptionAction = (id) => ({
  type: REMOVE_OPTION,
  payload: {
    id
  }
})

export const updateOptionAction = (id, data) => ({
  type: UPDATE_OPTION,
  payload: {
    id,
    data,
  }
})

export const updateOptionOrderAction = (order) => ({
  type: UPDATE_OPTION_ORDER,
  payload: {
    order
  }
})

export const addOption = (id) => (dispatch, getState) => {
  const { data } = getState().poll.options
  // remove any options that dont have text when adding a new one
  Object.entries(data).forEach(([id, option]) => {
    if (!option.text) {
      dispatch(removeOptionAction(id))
    } else {
      dispatch(updateOptionAction(id, { editing: false }))
    }
  })

  dispatch(addOptionAction(id))
}

export const updateOptionOrder = (order) => (dispatch) => {
  dispatch(updateOptionOrderAction(order))
}

export const removeOption = (id) => (dispatch) => {
  dispatch(removeOptionAction(id))
}

export const updateOption = (id, data) => (dispatch) => {
  dispatch(updateOptionAction(id, data))
}