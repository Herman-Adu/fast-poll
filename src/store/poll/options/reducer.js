import {
  ADD_OPTION,
  REMOVE_OPTION,
  UPDATE_OPTION,
  UPDATE_OPTION_ORDER,
  RESET_OPTIONS,
  UPDATE_NEWEST,
} from './actions';

export const INITIAL_STATE = {
  // [optionId]
  order: [],
  // options is a map with the optionId as the keys and
  // option data including results as the values
  data: {},
  newest: {
    id: null,
    text: '',
  },
};

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case ADD_OPTION:
      return {
        ...state,
        order: [...state.order, action.payload.id],
        data: {
          ...state.data,
          [action.payload.id]: {
            id: action.payload.id,
            editing: false,
            votes: 0,
            ...action.payload.data,
          },
        },
      };

    case REMOVE_OPTION:
      const { [action.payload.id]: removedOption, ...newOptions } = state.data;

      return {
        ...state,
        data: newOptions,
        order: state.order.filter(id => id !== action.payload.id),
      };

    case UPDATE_OPTION:
      return {
        ...state,
        data: {
          ...state.data,
          [action.payload.id]: {
            ...state.data[action.payload.id],
            ...Object.keys(action.payload.data).reduce((aggr, curr) => {
              if (curr === 'votes') {
                // eslint-disable-next-line no-param-reassign
                aggr.votes =
                  action.payload.id in state.data &&
                  'votes' in state.data[action.payload.id]
                    ? state.data[action.payload.id].votes +
                      action.payload.data.votes
                    : action.payload.data.votes;
              } else {
                // eslint-disable-next-line no-param-reassign
                aggr[curr] = action.payload.data[curr];
              }
              return aggr;
            }, {}),
          },
        },
      };

    case UPDATE_OPTION_ORDER:
      return {
        ...state,
        order: [...action.payload.order],
      };

    case UPDATE_NEWEST:
      return {
        ...state,
        newest: {
          ...state.newest,
          ...action.payload.data,
        },
      };

    case RESET_OPTIONS:
      return INITIAL_STATE;

    default:
      return state;
  }
};
