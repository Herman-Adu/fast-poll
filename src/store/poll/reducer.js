import { combineReducers } from 'redux';

import optionsReducer from './options/reducer';
import {
  LOADING,
  UPDATE_TITLE,
  UPDATE_SELECTION,
  SET_SHOW_RESULTS,
  SET_CREATED_POLL,
  SET_OWNER,
  RESET_POLL,
} from './actions';

export const INITIAL_STATE = {
  loading: false,
  title: '',
  selection: '',
  showResults: false,
  created: '',
  createdBy: '',
};

const rootReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case LOADING:
      return {
        ...state,
        loading: action.payload.isLoading,
      };

    case UPDATE_TITLE:
      return {
        ...state,
        title: action.payload.title,
      };

    case UPDATE_SELECTION:
      return {
        ...state,
        selection: action.payload.selection,
      };

    case SET_SHOW_RESULTS:
      return {
        ...state,
        showResults: action.payload.showResults,
      };

    case SET_CREATED_POLL:
      return {
        ...state,
        created: action.payload.id,
      };

    case SET_OWNER:
      return {
        ...state,
        createdBy: action.payload.createdBy,
      };

    case RESET_POLL:
      return INITIAL_STATE;

    default:
      return state;
  }
};

export default combineReducers({ data: rootReducer, options: optionsReducer });
