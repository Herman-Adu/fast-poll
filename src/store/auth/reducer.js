import { SIGN_IN, SIGN_OUT, LOADING } from './actions';

export const INITIAL_STATE = {
  loading: false,
  uid: '',
  isAnonymous: null,
  photoURL: '',
  // // some other properties from the user object that may be useful
  // email: '',
  // displayName: '',
};

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case SIGN_IN:
      return {
        ...state,
        ...action.payload,
      };

    case SIGN_OUT:
      return INITIAL_STATE;

    case LOADING:
      return {
        ...state,
        loading: action.payload.isLoading,
      };

    default:
      return state;
  }
};
