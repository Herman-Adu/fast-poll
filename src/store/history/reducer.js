import { LOADING, ADD_POLLS } from './actions';

export const INITIAL_STATE = {
  created: [],
  votedOn: [],
  loading: false,
};

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case LOADING:
      return {
        ...state,
        loading: action.payload.isLoading,
      };

    case ADD_POLLS:
      return {
        ...state,
        [action.payload.type]: [
          ...state[action.payload.type],
          ...action.payload.polls,
        ],
      };

    default:
      return state;
  }
};
