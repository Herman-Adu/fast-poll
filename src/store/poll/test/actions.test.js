import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { INITIAL_STATE } from '../reducer';
import * as actions from '../actions';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('poll bound actions', () => {
  it('setShowResults', () => {
    const store = mockStore(INITIAL_STATE);
    store.dispatch(actions.setShowResults(true));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('selectOption', () => {
    const store = mockStore(INITIAL_STATE);
    store.dispatch(actions.selectOption('1'));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('updateTitle', () => {
    const store = mockStore(INITIAL_STATE);
    store.dispatch(actions.updateTitle('new title'));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('setCreatedPoll', () => {
    const store = mockStore(INITIAL_STATE);
    store.dispatch(actions.setCreatedPoll('1'));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('resetPoll', () => {
    const store = mockStore(INITIAL_STATE);
    store.dispatch(actions.resetPoll());
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('poll async bound actions', () => {
  // TODO: figure out how to mock firebase
});
