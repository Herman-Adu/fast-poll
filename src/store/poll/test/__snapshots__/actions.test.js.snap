// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`poll bound actions resetPoll 1`] = `
Array [
  Object {
    "payload": null,
    "type": "poll/RESET_POLL",
  },
]
`;

exports[`poll bound actions selectOption 1`] = `
Array [
  Object {
    "payload": Object {
      "selection": "1",
    },
    "type": "poll/UPDATE_SELECTION",
  },
]
`;

exports[`poll bound actions setCreatedPoll 1`] = `
Array [
  Object {
    "payload": Object {
      "id": "1",
    },
    "type": "poll/SET_CREATED_POLL",
  },
]
`;

exports[`poll bound actions setShowResults 1`] = `
Array [
  Object {
    "payload": Object {
      "showResults": true,
    },
    "type": "poll/SET_SHOW_RESULTS",
  },
]
`;

exports[`poll bound actions updateTitle 1`] = `
Array [
  Object {
    "payload": Object {
      "title": "new title",
    },
    "type": "poll/UPDATE_TITLE",
  },
]
`;
